### PR TITLE
Payment Checkout: implement redirect modal in the checkout page for Agency/Pro users from Jetpack Pricing or My Jetpack page

### PR DESCRIPTION
**Pull request by yashwin**
Related to 1202619025189113-as-1204953498074085

## Proposed Changes

This PR implements a modal in the payment checkout page that shows the Agency/Pro users the benefits of buying the products from the Jetpack Licensing Portal. 

We show this page only if 

- The user is an agency partner.
- If the user is coming from Jetpack Pricing or My Jetpack page.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/show-redirect-modal-in-checkout` and `yarn start` or open the Calypso live link. 
2. Open http://calypso.localhost:3000/ or the Calypso live link.
3. Visit https://cloud.jetpack.com/pricing and click on the `Get` button on any product -> You'll be redirected to the payment checkout page -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL.
4. Verify that you can see the modal as shown below

mobile (23)

5. Go to WP.com -> Select any Jetpack site -> Visit WP Admin page -> Click on Jetpack - My Jetpack -> Click on `Start for free` on the Search card -> Click the `Get Jetpack Search` button -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL and verify you can now see the same modal.

6. Go to WP.com -> Select any Jetpack site -> Visit WP Admin page -> Click on Jetpack - My Jetpack -> Click on `Purchase` on any product(Backup or Scan) -> Click the get product button -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL and verify you can now see the same modal.
7. Verify the modal looks good on different screen sizes

**Tablet**

mobile (24)

**Mobile**

mobile (25)

8. Go to WP.com and add any WP.com plans(Personal or Business) by clicking the Upgrade button and verify that the modal is not shown now.

Screenshot 2023-07-05 at 10 11 14 AM

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by simonktnga8c** [comment]
Just tested and it works great; nice work @yashwin!

**Review by ilonagl — APPROVED** [review]
Looks great. Amazing work!

File: client/my-sites/checkout/composite-checkout/components/checkout-main.tsx [hunk 1]
```diff
@@ -722,26 +728,34 @@ export default function CheckoutMain( {
...
-					createUserAndSiteBeforeTransaction={ createUserAndSiteBeforeTransaction }
-					infoMessage={ infoMessage }
-					isLoggedOutCart={ !! isLoggedOutCart }
-					onPageLoadError={ onPageLoadError }
-					removeProductFromCart={ removeProductFromCartAndMaybeRedirect }
-					showErrorMessageBriefly={ showErrorMessageBriefly }
-					siteId={ updatedSiteId }
-					siteUrl={ updatedSiteSlug }
-				/>
+				<>
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/components/checkout-main.tsx]: In theory, you should be able to have multiple sibling children of `CheckoutProvider` without needing to add this fragment, but I'm guessing you did it because there was a problem. If that's so, what was the issue that this solved?

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import { Button } from '@automattic/components';
+import { Modal } from '@wordpress/components';
+import { useTranslate } from 'i18n-calypso';
+import { CheckIcon } from 'calypso/../packages/composite-checkout/src/components/shared-icons';
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: This should be a package import. We should not reach inside the private implementation details of a package. 

```suggestion
import { CheckIcon } from '@automattic/composite-checkout';
```

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+							{ translate( 'Get a recurring discount and more flexible billing' ) }
+						</div>
+						<p>
+							{ translate(
+								'As a Jetpack partner, by purchasing products through our Agency & Pro tools, you get:'
+							) }
+						</p>
+						<ul className="jetpack-pro-redirect-modal__list">
+							{ features.map( ( feature, index ) => (
+								<li key={ index }>
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: [You should not use the index as the key](https://react.dev/learn/rendering-lists#why-does-react-need-keys). There's a ton of other articles about why this is dangerous out there on the web if you need additional info. Can we use `feature` here?

**Review by sirbrillig — CHANGES_REQUESTED** [review]
Thank you for providing screenshots at different widths! Just a few comments.

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 2 above)
Comment (yashwin) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: I think this happened because of auto-import. I agree with this and fixed it here - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

File: client/my-sites/checkout/composite-checkout/components/checkout-main.tsx (hunk 1 above)
Comment (yashwin) [reply client/my-sites/checkout/composite-checkout/components/checkout-main.tsx]: Yeah. It did work - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 3 above)
Comment (yashwin) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: Totally agree, and sorry for missing this. I have fixed it now - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

**Comment by yashwin** [comment]
@sirbrillig: Thanks for your thorough review and feedback. I have addressed all your comments. Could you please take another look?

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 2 above)
Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: Yeah, that happens to me all the time too. I should check with the Calypso team to see if there's a way to add a rule to catch these sorts of imports. 🤔 @jsnajdr do you think that would be possible and/or would it be too much work?

**Review by sirbrillig — APPROVED** [review]
I haven't run through the testing instructions but from a payments perspective, this looks good to me. Thanks for the fixes!

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	// Show the banner only if the user has not dismissed the banner and
+	// is coming from the Jetpack Plans page or Jetpack page in WP.com
+	if (
+		isDismissed ||
+		! (
+			productSourceFromUrl === 'jetpack-plans' ||
+			[ 'my-jetpack', 'jetpack' ].includes( redirectURLPage )
+		)
+	) {
+		return null;
```
Comment (jkguidaven) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:62]: Not a blocker.

I see two separate places where we check if we need to display this modal. What do you think about moving this logic to the `checkout-main.tsx` or adding the Agency user check here?

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return (
+		<>
```
Comment (jkguidaven) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: No need to wrap the Modal with `<>`.

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/style.scss [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		display: flex;
+		justify-content: flex-start;
+		flex-direction: column;
+		position: absolute;
+		bottom: 0;
+		left: 0;
+		padding: 16px 32px;
+		background: var(--studio-white);
+
+		.jetpack-pro-redirect-modal__footer-buttons {
```
Comment (jkguidaven) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/style.scss:92]: Just nitpicking but I think we can shorten this to `&-buttons`.

File: client/state/jetpack-agency-dashboard/selectors.ts [hunk 7]
```diff
@@ -16,6 +16,9 @@ export const JETPACK_DASHBOARD_WELCOME_BANNER_PREFERENCE_HOME_PAGE =
 export const JETPACK_DASHBOARD_SURVEY_BANNER_PREFERENCE =
 	'jetpack-dashboard-agency-program-survey-banner-preference';
 
+export const JETPACK_DASHBOARD_CHECKOUT_REDIRECT_MODAL_DISMISSED =
+	'jetpack-dashboard-checkout-redirect-modal-dismissed';
```
Comment (jkguidaven) [inline client/state/jetpack-agency-dashboard/selectors.ts]: Not a blocker.

Maybe we can add the word `agency-program` in the preference name just to make it more specific to the agency program and not overlap with the Jetpack dashboard in the core plugin. wdyt?

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			<Modal
+				open={ true }
+				onRequestClose={ dismissAndRecordEvent }
+				title=""
```
Comment (jkguidaven) [inline client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: Maybe what we need here is **Dialog** instead of **Modal**.
```
import { Dialog } from '@automattic/components';
```

**Review by jkguidaven — COMMENTED** [review]
Tested it and it is working as expected. Nice work @yashwin 💯 👍 

However, I left few comments.

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/style.scss (hunk 6 above)
Comment (yashwin) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/style.scss:92]: I would like to keep it like this, keeping searchability in mind.

File: client/state/jetpack-agency-dashboard/selectors.ts (hunk 7 above)
Comment (yashwin) [reply client/state/jetpack-agency-dashboard/selectors.ts]: Doesn't really matter. But, I have changed this.

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 8 above)
Comment (yashwin) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: I'm not going to make this change here as it requires a lot of changes.Also, there is not much difference between using these 2 components.

**Comment by yashwin** [comment]
@jkguidaven - Thanks for your thorough review. I have addressed all your comments and made the changes - https://github.com/Automattic/wp-calypso/pull/78955/commits/7be78d89fc5eaf2cc2f65f6493f323ce01141012

We would want to verify this for non-agency users, as we have moved the conditions to one place. This modal should ideally not appear to non-agency users.

**Comment by jkguidaven** [comment]
Thanks for addressing all my PR comments. Tested both agency and non-agency user flow, and it work as expected. this LGTM 👍 🚢

**Review by jkguidaven — APPROVED** [review]

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 2 above)
Comment (jsnajdr) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: > see if there's a way to add a rule to catch these sorts of imports.

You mean an eslint rule? Maybe `eslint-plugin-import` has some rule for this out of the box. If we wanted to write our own rule, it shouldn't be terribly hard. We'd mainly have to specify what exactly are the imports we want to avoid.

The auto-import suggestions are created by VSCode, which in turn uses the JS Language Server to generate them.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8359905

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8359905).

File: client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx (hunk 2 above)
Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx]: >  We'd mainly have to specify what exactly are the imports we want to avoid.

I think that it would be safe to say that we want to block all imports that try to reach a path starting with `calypso/../packages`. To also block relative imports we could say any path including `../packages` but it is possible that could find false positives.
